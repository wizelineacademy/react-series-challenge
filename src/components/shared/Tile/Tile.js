import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { bindActionCreators } from "redux";

import constants from "./TileConstants";
import ImageLoader from "../ImageLoader";
import actions from "../../Favs/FavsActions";
import { TileWrapperStyled, FavoriteMarkStyled } from "./Tile.styled";

const Tile = ({ gif, original, favs, loaded, toggleFavorite }) => {
  const dimensions = original ? gif.images.original : constants;
  const { width, height } = dimensions;

  gif.favorite = favs.some(fav => fav.id === gif.id);

  const handleClick = event => {
    event.preventDefault();
    toggleFavorite(gif);
  };

  return (
    <Link to={`/details/${gif.id}`}>
      <TileWrapperStyled width={width} height={height}>
        {loaded[gif.id] && (
          <FavoriteMarkStyled marked={gif.favorite} onClick={handleClick}>
            ‚ù§
          </FavoriteMarkStyled>
        )}
        <ImageLoader
          image={gif}
          width={width}
          height={height}
          original={original}
        />
      </TileWrapperStyled>
    </Link>
  );
};

const mapStateToProps = state => {
  const { favs } = state.favsReducer;
  const { loaded } = state.imageLoaderReducer;
  return { favs, loaded };
};

const mapDispatchToProps = dispatch => {
  const { toggleFavorite } = actions.creators;
  return bindActionCreators({ toggleFavorite }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(Tile);
