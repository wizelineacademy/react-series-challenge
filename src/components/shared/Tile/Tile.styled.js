import styled, { keyframes } from "styled-components";

const animation = ({ width }) => keyframes`
  0% { background-position: -${width}px 0; }
  100% { background-position: ${width}px 0; }
`;

const TileWrapperStyled = styled.div`
  width: ${props => props.width}px;
  height: ${props => props.height}px;
  position: relative;
  margin: 5px;
  transition: 0.1s ease-in-out transform;
  border-radius: 2px;
  background: linear-gradient(to right, #eeeeee 8%, #dddddd 18%, #eeeeee 33%);
  animation: 1.5s ${props => animation(props)} linear infinite;
  animation-fill-mode: forwards;
  background-size: 800px 104px;
  position: relative;

  &:hover {
    transform: scale(1.05);
  }
`;

const FavoriteMarkStyled = styled.label`
  z-index: 1;
  position: absolute;
  right: 14px;
  top: 8px;
  font-size: 2em;
  color: ${props => (props.marked ? "red" : "white")};
  opacity: ${props => (props.marked ? 1 : 0.4)};
  transition: 0.1s ease-in;
  user-select: none;

  &:hover {
    color: "red";
    cursor: pointer;
  }

  &:active {
    opacity: 1;
    transform: scale(1.1);
  }
`;

export { TileWrapperStyled, FavoriteMarkStyled };
