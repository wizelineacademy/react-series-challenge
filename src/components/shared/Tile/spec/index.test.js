import React from "react";
import { Provider } from "react-redux";
import { MemoryRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import "jest-styled-components";
import Tile from "../";

const mockStore = configureStore();

const shape = {
  gif: {
    id: 1,
    images: {
      original: { width: 400, height: 400, url: "http://original" },
      fixed_width: { width: 300, height: 300, url: "http://fixed" }
    }
  },
  original: true,
  color: "red"
};

const initialState = {
  favsReducer: { favs: [] },
  imageLoaderReducer: { loaded: {} }
};

describe("<Tile />", () => {
  it("renders properly", () => {
    const init = Object.assign(initialState);
    const store = mockStore(init);
    const props = Object.assign(shape);
    const component = renderer.create(
      <Provider store={store}>
        <Router>
          <Tile {...props} />
        </Router>
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders the heart mark", () => {
    const init = Object.assign(initialState);
    init.imageLoaderReducer.loaded = { 1: true };
    const store = mockStore(init);
    const props = Object.assign(shape);
    const component = renderer.create(
      <Provider store={store}>
        <Router>
          <Tile {...props} />
        </Router>
      </Provider>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("render the original image", () => {
    const init = Object.assign(initialState);
    const store = mockStore(init);
    const props = Object.assign(shape);
    const component = renderer.create(
      <Provider store={store}>
        <Router>
          <Tile {...props} />
        </Router>
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("render the small image", () => {
    const init = Object.assign(initialState);
    const store = mockStore(init);
    const props = Object.assign(shape);
    props.original = false;
    const component = renderer.create(
      <Provider store={store}>
        <Router>
          <Tile {...props} />
        </Router>
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("mark a gif as favorite before rendering", () => {
    const initialState = {
      favsReducer: { favs: [{ ...shape.gif }] },
      imageLoaderReducer: { loaded: { 1: true } }
    };
    const init = Object.assign(initialState);
    const store = mockStore(init);
    const props = Object.assign(shape);
    const component = renderer.create(
      <Provider store={store}>
        <Router>
          <Tile {...props} />
        </Router>
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
