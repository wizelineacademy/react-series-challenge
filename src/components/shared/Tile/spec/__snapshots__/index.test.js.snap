// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile /> mark a gif as favorite before rendering 1`] = `
.c2 {
  width: 300px;
  height: 300px;
  border-radius: 2px;
  object-fit: cover;
}

.c0 {
  width: 300px;
  height: 300px;
  position: relative;
  margin: 5px;
  -webkit-transition: 0.1s ease-in-out -webkit-transform;
  -webkit-transition: 0.1s ease-in-out transform;
  transition: 0.1s ease-in-out transform;
  border-radius: 2px;
  background: linear-gradient(to right,#eeeeee 8%,#dddddd 18%,#eeeeee 33%);
  -webkit-animation: 1.5s eNVqyK linear infinite;
  animation: 1.5s eNVqyK linear infinite;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  background-size: 800px 104px;
  position: relative;
}

.c0:hover {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c1 {
  z-index: 1;
  position: absolute;
  right: 14px;
  top: 8px;
  font-size: 2em;
  color: red;
  opacity: 1;
  -webkit-transition: 0.1s ease-in;
  transition: 0.1s ease-in;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1:hover {
  color: "red";
  cursor: pointer;
}

.c1:active {
  opacity: 1;
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

<a
  href="/details/1"
  onClick={[Function]}
>
  <div
    className="c0"
    height={300}
    width={300}
  >
    <label
      className="c1"
      onClick={[Function]}
    >
      ❤
    </label>
    <img
      className="c2"
      height={300}
      src={undefined}
      width={300}
    />
  </div>
</a>
`;

exports[`<Tile /> render the original image 1`] = `
.c2 {
  width: 400px;
  height: 400px;
  border-radius: 2px;
  object-fit: cover;
}

.c0 {
  width: 400px;
  height: 400px;
  position: relative;
  margin: 5px;
  -webkit-transition: 0.1s ease-in-out -webkit-transform;
  -webkit-transition: 0.1s ease-in-out transform;
  transition: 0.1s ease-in-out transform;
  border-radius: 2px;
  background: linear-gradient(to right,#eeeeee 8%,#dddddd 18%,#eeeeee 33%);
  -webkit-animation: 1.5s bFrnLp linear infinite;
  animation: 1.5s bFrnLp linear infinite;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  background-size: 800px 104px;
  position: relative;
}

.c0:hover {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c1 {
  z-index: 1;
  position: absolute;
  right: 14px;
  top: 8px;
  font-size: 2em;
  color: white;
  opacity: 0.4;
  -webkit-transition: 0.1s ease-in;
  transition: 0.1s ease-in;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1:hover {
  color: "red";
  cursor: pointer;
}

.c1:active {
  opacity: 1;
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

<a
  href="/details/1"
  onClick={[Function]}
>
  <div
    className="c0"
    height={400}
    width={400}
  >
    <label
      className="c1"
      onClick={[Function]}
    >
      ❤
    </label>
    <img
      className="c2"
      height={400}
      src={undefined}
      width={400}
    />
  </div>
</a>
`;

exports[`<Tile /> render the small image 1`] = `
.c2 {
  width: 300px;
  height: 300px;
  border-radius: 2px;
  object-fit: cover;
}

.c0 {
  width: 300px;
  height: 300px;
  position: relative;
  margin: 5px;
  -webkit-transition: 0.1s ease-in-out -webkit-transform;
  -webkit-transition: 0.1s ease-in-out transform;
  transition: 0.1s ease-in-out transform;
  border-radius: 2px;
  background: linear-gradient(to right,#eeeeee 8%,#dddddd 18%,#eeeeee 33%);
  -webkit-animation: 1.5s eNVqyK linear infinite;
  animation: 1.5s eNVqyK linear infinite;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  background-size: 800px 104px;
  position: relative;
}

.c0:hover {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c1 {
  z-index: 1;
  position: absolute;
  right: 14px;
  top: 8px;
  font-size: 2em;
  color: white;
  opacity: 0.4;
  -webkit-transition: 0.1s ease-in;
  transition: 0.1s ease-in;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1:hover {
  color: "red";
  cursor: pointer;
}

.c1:active {
  opacity: 1;
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

<a
  href="/details/1"
  onClick={[Function]}
>
  <div
    className="c0"
    height={300}
    width={300}
  >
    <label
      className="c1"
      onClick={[Function]}
    >
      ❤
    </label>
    <img
      className="c2"
      height={300}
      src={undefined}
      width={300}
    />
  </div>
</a>
`;

exports[`<Tile /> renders properly 1`] = `
.c0 {
  width: 400px;
  height: 400px;
  position: relative;
  margin: 5px;
  -webkit-transition: 0.1s ease-in-out -webkit-transform;
  -webkit-transition: 0.1s ease-in-out transform;
  transition: 0.1s ease-in-out transform;
  border-radius: 2px;
  background: linear-gradient(to right,#eeeeee 8%,#dddddd 18%,#eeeeee 33%);
  -webkit-animation: 1.5s bFrnLp linear infinite;
  animation: 1.5s bFrnLp linear infinite;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  background-size: 800px 104px;
  position: relative;
}

.c0:hover {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

<a
  href="/details/1"
  onClick={[Function]}
>
  <div
    className="c0"
    height={400}
    width={400}
  />
</a>
`;

exports[`<Tile /> renders the heart mark 1`] = `
.c2 {
  width: 400px;
  height: 400px;
  border-radius: 2px;
  object-fit: cover;
}

.c0 {
  width: 400px;
  height: 400px;
  position: relative;
  margin: 5px;
  -webkit-transition: 0.1s ease-in-out -webkit-transform;
  -webkit-transition: 0.1s ease-in-out transform;
  transition: 0.1s ease-in-out transform;
  border-radius: 2px;
  background: linear-gradient(to right,#eeeeee 8%,#dddddd 18%,#eeeeee 33%);
  -webkit-animation: 1.5s bFrnLp linear infinite;
  animation: 1.5s bFrnLp linear infinite;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  background-size: 800px 104px;
  position: relative;
}

.c0:hover {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c1 {
  z-index: 1;
  position: absolute;
  right: 14px;
  top: 8px;
  font-size: 2em;
  color: white;
  opacity: 0.4;
  -webkit-transition: 0.1s ease-in;
  transition: 0.1s ease-in;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1:hover {
  color: "red";
  cursor: pointer;
}

.c1:active {
  opacity: 1;
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

<a
  href="/details/1"
  onClick={[Function]}
>
  <div
    className="c0"
    height={400}
    width={400}
  >
    <label
      className="c1"
      onClick={[Function]}
    >
      ❤
    </label>
    <img
      className="c2"
      height={400}
      src={undefined}
      width={400}
    />
  </div>
</a>
`;
